src/vmm/src/pipe.rs

```diff
@@ -0,0 +1,69 @@
+// Copyright 2020 Sartura All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+#![cfg(target_arch = "x86_64")]
+
+use std::io::{Read, Write};
+
+use std::fs::{File};
+use std::os::unix::io::{AsRawFd, RawFd};
+
+use nix::sys::stat;
+use nix::unistd;
+use std::path::PathBuf;
+
+/// Newtype for piping the output of guest cmd.
+pub(crate) struct PipeReaderWrapper(PathBuf, File);
+
+impl PipeReaderWrapper {
+    pub fn new(buf: PathBuf) -> PipeReaderWrapper {
+        match unistd::mkfifo(&buf, stat::Mode::S_IRWXO) {
+            Ok(_) => (),
+            Err(err) => panic!("Error creating fifo: {}", err),
+        }
+        let pipe = match File::open(&buf) {
+            Err(err) => panic!("couldn't open pipe: {}", err),
+            Ok(output) => output,
+        };
+        return PipeReaderWrapper(buf, pipe);
+    }
+}
+
+impl Read for PipeReaderWrapper {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        return self.1.read(buf);
+    }
+}
+
+impl AsRawFd for PipeReaderWrapper {
+    fn as_raw_fd(&self) -> RawFd {
+        return self.1.as_raw_fd();
+    }
+}
+
+pub(crate) struct PipeWriterWrapper(PathBuf);
+
+impl PipeWriterWrapper {
+    pub fn new(buf: PathBuf) -> PipeWriterWrapper {
+        match unistd::mkfifo(&buf, stat::Mode::S_IRWXO) {
+            Ok(_) => (),
+            Err(err) => panic!("Error creating fifo: {}", err),
+        }
+        return PipeWriterWrapper(buf);
+    }
+}
+
+impl Write for PipeWriterWrapper {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        let mut pipe = match File::create(&self.0) {
```

Comment:
I think this needs `OpenOptions` to make sure it's writable and to control the opening behavior (`O_CREAT`, `O_TRUNC` etc).